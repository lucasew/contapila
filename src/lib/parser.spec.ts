import { describe, test, expect, beforeEach } from 'vitest';
import {
  REGEX_PATTERNS,
  createParserConfig,
  createParser,
  parseAmount,
  parseTag,
  parseTags,
  convertBeancountToGeneralizedFormat,
  validateModuleDependencies
} from './parser.js';
import { createCoreBeancountModule } from './beancount.js';

// Utilitário de teste para criar um cursor
const createTestCursor = (text: string, position = 0) => ({
  text,
  position,
  line: 1,
  column: 1
});

// Os testes serão movidos do beancount.spec.ts para cá. 

describe('Regex Patterns', () => {
	test('DATE pattern matches valid dates', () => {
		expect('2024-01-01'.match(REGEX_PATTERNS.DATE)).toBeTruthy();
		expect('2024-12-31'.match(REGEX_PATTERNS.DATE)).toBeTruthy();
		expect('invalid-date'.match(REGEX_PATTERNS.DATE)).toBeFalsy();
		expect('24-01-01'.match(REGEX_PATTERNS.DATE)).toBeFalsy();
	});

	test('ACCOUNT pattern matches valid accounts', () => {
		expect('Assets:Cash'.match(REGEX_PATTERNS.ACCOUNT)).toBeTruthy();
		expect('Expenses:Food:Groceries'.match(REGEX_PATTERNS.ACCOUNT)).toBeTruthy();
		expect('Income:Salary_2024'.match(REGEX_PATTERNS.ACCOUNT)).toBeTruthy();
		expect('assets:cash'.match(REGEX_PATTERNS.ACCOUNT)).toBeFalsy(); // lowercase first letter
	});
});

describe('Core Parser Functions', () => {
	describe('Cursor utilities', () => {
		test('createCursor initializes corretamente', () => {
			const cursor = createTestCursor('test text');
			expect(cursor.text).toBe('test text');
			expect(cursor.position).toBe(0);
			expect(cursor.line).toBe(1);
			expect(cursor.column).toBe(1);
		});
	});

	describe('Basic parsers', () => {
		test('parseAmount extrai valores com moedas', () => {
			const config = createParserConfig([]);
			expect(parseAmount(createTestCursor('100.50 USD'))?.value).toStrictEqual({
				value: 100.5,
				currency: 'USD'
			});
			expect(parseAmount(createTestCursor('-200.75 BRL'))?.value).toStrictEqual({
				value: -200.75,
				currency: 'BRL'
			});
			expect(parseAmount(createTestCursor('1000EUR'))?.value).toStrictEqual({
				value: 1000,
				currency: 'EUR'
			});
			expect(parseAmount(createTestCursor('1000A_B_C'))?.value).toStrictEqual({
				value: 1000,
				currency: 'A_B_C'
			});
			expect(parseAmount(createTestCursor('invalid amount'))).toBeNull();
		});

		test('parseTag funciona diretamente', () => {
			const cursor = createTestCursor('#test');
			const result = parseTag(cursor);
			expect(result).not.toBeNull();
			expect(result?.value).toBe('test');
		});

		test('parseTags funciona diretamente', () => {
			const cursor = createTestCursor('#tag1 #tag2');
			const result = parseTags(cursor);
			expect(result).not.toBeNull();
			expect(result?.value).toEqual(['tag1', 'tag2']);
		});
	});
});

describe('YAML Parser', () => {
	test('parses simple key-value pairs', () => {
		const config = createParserConfig([createCoreBeancountModule()]);
		const parser = createParser(config);

		const entryWithMeta = `2024-01-01 open Assets:Cash USD\n  description: \"Main account\"\n  priority: high\n  active: true\n  balance: 1000.50`;

		const entries = parser(entryWithMeta);

		expect(entries).toHaveLength(1);
		expect(entries[0].meta).toMatchObject({
			description: 'Main account',
			priority: 'high',
			active: true,
			balance: 1000.5
		});
		expect(typeof entries[0].meta.location).toBe('string');
		expect(entries[0].meta.location).toMatch(/^(\$file|stdin):\d+$/);
	});

	test('handles different data types in YAML', () => {
		const config = createParserConfig([createCoreBeancountModule()]);
		const parser = createParser(config);

		const entryWithMeta = `2024-01-01 open Assets:Cash USD\n  string_value: test\n  integer_value: 42\n  float_value: 3.14\n  boolean_true: true\n  boolean_false: false\n  null_value: `;

		const entries = parser(entryWithMeta);

		expect(entries[0].meta).toMatchObject({
			string_value: 'test',
			integer_value: 42,
			float_value: 3.14,
			boolean_true: true,
			boolean_false: false,
			null_value: null
		});
		expect(typeof entries[0].meta.location).toBe('string');
		expect(entries[0].meta.location).toMatch(/^(\$file|stdin):\d+$/);
	});
});

describe('Funções utilitárias do parser', () => {
	test('convertBeancountToGeneralizedFormat transforma corretamente', () => {
		const beancountEntries = [
			{
				type: 'transaction',
				date: '2024-01-01',
				flag: '*',
				payee: 'Store',
				narration: 'Purchase',
				postings: [
					{
						account: 'Assets:Cash',
						amount: { number: -100, currency: 'USD' },
						cost: { number: -100, currency: 'USD' },
						price: { number: 1, currency: 'USD' },
						flag: '*',
						meta: { note: 'test' }
					}
				],
				meta: { category: 'shopping' }
			},
			{
				type: 'balance',
				date: '2024-01-01',
				account: 'Assets:Cash',
				amount: { number: 1000, currency: 'USD' },
				meta: { verified: true }
			}
		];

		const config = createParserConfig([]);
		const converted = convertBeancountToGeneralizedFormat(beancountEntries, config);

		expect(converted).toHaveLength(2);

		expect(converted[0]).toMatchObject({
			kind: 'transaction',
			date: '2024-01-01',
			flag: '*',
			payee: 'Store',
			narration: 'Purchase',
			meta: { category: 'shopping' }
		});

		expect(converted[0].postings[0]).toMatchObject({
			account: 'Assets:Cash',
			amount: { value: -100, currency: 'USD' },
			cost: { value: -100, currency: 'USD' },
			price: { value: 1, currency: 'USD' },
			flag: '*',
			meta: { note: 'test' }
		});

		expect(converted[1]).toMatchObject({
			kind: 'balance',
			date: '2024-01-01',
			account: 'Assets:Cash',
			amount: { value: 1000, currency: 'USD' },
			meta: { verified: true }
		});
	});

	test('validateModuleDependencies detecta dependências faltantes', () => {
		const modules: any[] = [
			{
				name: 'module-a',
				version: '1.0.0',
				dependencies: ['missing-module'],
				directives: []
			}
		];

		const errors = validateModuleDependencies(modules);
		expect(errors).toHaveLength(1);
		expect(errors[0]).toContain("depends on missing module 'missing-module'");
	});

	test('validateModuleDependencies passa com dependências válidas', () => {
		const modules: any[] = [
			{
				name: 'core',
				version: '1.0.0',
				directives: []
			},
			{
				name: 'extended',
				version: '1.0.0',
				dependencies: ['core'],
				directives: []
			}
		];

		const errors = validateModuleDependencies(modules);
		expect(errors).toHaveLength(0);
	});
});

describe('Linha com ;; e string', () => {
	test('parses open directive with double semicolon and quoted string', () => {
		const config = createParserConfig([createCoreBeancountModule()]);
		const parser = createParser(config);
		const text = '2015-03-01 open Assets:XXX:YYY ;; "SOME_STRING"';
		const entries = parser(text);
		expect(entries).toHaveLength(1);
		expect(entries[0]).toMatchObject({
			kind: 'open',
			date: '2015-03-01',
			account: 'Assets:XXX:YYY',
		});
	});
}); 