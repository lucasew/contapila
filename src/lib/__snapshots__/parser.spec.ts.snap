// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`INTEGRAÇÃO: exemplo completo do arquivo example.beancount > parseia todas as entidades do exemplo exatamente como esperado 1`] = `
[
  {
    "account": "Assets:Cash",
    "currencies": [
      "USD",
    ],
    "date": "2015-01-01",
    "keyword": "open",
    "kind": "open",
    "links": [
      "init",
    ],
    "meta": {
      "description": "Conta principal",
      "location": "example.beancount:3",
      "prioridade": "alta",
    },
    "tags": [
      "start",
    ],
  },
  {
    "account": "Assets:Investments:Stocks",
    "date": "2015-01-01",
    "keyword": "open",
    "kind": "open",
    "meta": {
      "ativo": "PETR4",
      "corretora": "XP",
      "location": "example.beancount:7",
      "valor_inicial": 10000,
    },
  },
  {
    "account": "Expenses:Food",
    "date": "2015-01-01",
    "keyword": "open",
    "kind": "open",
    "links": [
      "supermercado",
    ],
    "meta": {
      "location": "example.beancount:12",
    },
    "tags": [
      "grocery",
    ],
  },
  {
    "account": "Expenses:Food",
    "date": "2015-01-01",
    "keyword": "close",
    "kind": "close",
    "links": [
      "fim",
    ],
    "meta": {
      "location": "example.beancount:14",
    },
    "tags": [
      "cleanup",
    ],
  },
  {
    "account": "Assets:Cash",
    "amount": {
      "currency": "USD",
      "value": 1000,
    },
    "date": "2015-01-15",
    "keyword": "balance",
    "kind": "balance",
    "links": [
      "saldo",
    ],
    "meta": {
      "location": "example.beancount:16",
      "verificado": true,
    },
    "tags": [
      "monthly",
    ],
  },
  {
    "amount": {
      "currency": "BRL",
      "value": 5.25,
    },
    "commodity": "USD",
    "date": "2015-01-20",
    "keyword": "price",
    "kind": "price",
    "links": [
      "usdbrl",
    ],
    "meta": {
      "location": "example.beancount:19",
    },
    "tags": [
      "fx",
    ],
  },
  {
    "account": "Assets:Cash",
    "comment": "Primeira nota",
    "date": "2015-02-01",
    "keyword": "note",
    "kind": "note",
    "links": [
      "notinha",
    ],
    "meta": {
      "location": "example.beancount:21",
    },
    "tags": [
      "info",
    ],
  },
  {
    "date": "2015-02-10",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "compra",
    ],
    "meta": {
      "categoria": "alimentação",
      "recibo": "12345",
    },
    "narration": "Compra semanal",
    "payee": "Supermercado",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": -150.75,
        },
        "meta": undefined,
      },
      {
        "account": "Expenses:Food",
        "amount": {
          "currency": "USD",
          "value": 150.75,
        },
        "meta": {
          "tax_included": true,
        },
      },
    ],
    "tags": [
      "food",
    ],
  },
  {
    "date": "2015-02-15",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "restaurante",
    ],
    "meta": undefined,
    "narration": "Almoço",
    "payee": "Restaurante",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": -50,
        },
        "meta": undefined,
      },
      {
        "account": "Expenses:Food",
        "amount": {
          "currency": "USD",
          "value": 50,
        },
        "meta": undefined,
      },
    ],
    "tags": [
      "food",
    ],
  },
  {
    "date": "2015-03-01",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "salario",
    ],
    "meta": undefined,
    "narration": "Recebimento",
    "payee": "Salário",
    "postings": [
      {
        "account": "Income:Salary",
        "amount": {
          "currency": "USD",
          "value": -3000,
        },
        "meta": undefined,
      },
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": 3000,
        },
        "meta": undefined,
      },
    ],
    "tags": [
      "income",
    ],
  },
  {
    "date": "2015-03-05",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "b3",
    ],
    "meta": undefined,
    "narration": "Compra de ação",
    "payee": "Investimento",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": -500,
        },
        "meta": undefined,
      },
      {
        "account": "Assets:Investments:Stocks",
        "amount": {
          "currency": "USD",
          "value": 500,
        },
        "meta": {
          "ativo": "PETR4",
          "quantidade": 10,
        },
      },
    ],
    "tags": [
      "invest",
    ],
  },
  {
    "date": "2015-03-10",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "move",
    ],
    "meta": undefined,
    "narration": "Entre contas",
    "payee": "Transferência",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": -200,
        },
        "meta": undefined,
      },
      {
        "account": "Assets:Bank",
        "amount": {
          "currency": "USD",
          "value": 200,
        },
        "meta": undefined,
      },
    ],
    "tags": [
      "transfer",
    ],
  },
  {
    "date": "2015-03-15",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "amazon",
    ],
    "meta": undefined,
    "narration": "Amazon",
    "payee": "Compra internacional",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": -100,
        },
        "meta": undefined,
      },
      {
        "account": "Expenses:Shopping",
        "amount": {
          "currency": "USD",
          "value": 100,
        },
        "meta": {
          "moeda": "USD",
          "pais": "EUA",
        },
      },
    ],
    "tags": [
      "shopping",
    ],
  },
  {
    "date": "2015-03-20",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "zero",
    ],
    "meta": undefined,
    "narration": "Com valor zero",
    "payee": "Pagamento parcial",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": 0,
        },
        "meta": undefined,
      },
      {
        "account": "Expenses:Other",
        "amount": {
          "currency": "USD",
          "value": 0,
        },
        "meta": undefined,
      },
    ],
    "tags": [
      "weird",
    ],
  },
  {
    "date": "2015-03-25",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "linha",
    ],
    "meta": undefined,
    "narration": "Linha com comentário",
    "payee": "Comentário",
    "postings": [],
    "tags": [
      "comment",
    ],
  },
  {
    "body": "Assets:Cash      -10.00 USD",
    "date": "",
    "kind": "unknown_directive",
    "meta": {
      "body": "Assets:Cash      -10.00 USD",
      "location": "example.beancount:59",
      "type": "Assets:Cash",
      "warning": "Unknown directive at line 59, column 30",
    },
  },
  {
    "body": "Expenses:Other   10.00 USD",
    "date": "",
    "kind": "unknown_directive",
    "meta": {
      "body": "Expenses:Other   10.00 USD",
      "location": "example.beancount:60",
      "type": "Expenses:Other",
      "warning": "Unknown directive at line 60, column 29",
    },
  },
  {
    "body": "2015-04-01 unknown_directive Algo estranho aqui #unknown ^edge",
    "date": "",
    "kind": "unknown_directive",
    "meta": {
      "body": "2015-04-01 unknown_directive Algo estranho aqui #unknown ^edge",
      "location": "example.beancount:62",
      "type": "2015-04-01",
      "warning": "Unknown directive at line 62, column 63",
    },
  },
  {
    "body": "campo: "valor"",
    "date": "",
    "kind": "unknown_directive",
    "meta": {
      "body": "campo: "valor"",
      "location": "example.beancount:63",
      "type": "campo:",
      "warning": "Unknown directive at line 63, column 17",
    },
  },
  {
    "account": "Assets:Crypto:BTC",
    "date": "2015-04-10",
    "keyword": "open",
    "kind": "open",
    "meta": {
      "address": "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa",
      "exchange": "Binance",
      "location": "example.beancount:65",
    },
  },
  {
    "date": "2015-04-15",
    "flag": "*",
    "kind": "transaction",
    "links": undefined,
    "meta": undefined,
    "narration": "Teste de parser",
    "payee": "Transação com ;;",
    "postings": [],
    "tags": undefined,
  },
  {
    "body": "Assets:Cash      -1.00 USD",
    "date": "",
    "kind": "unknown_directive",
    "meta": {
      "body": "Assets:Cash      -1.00 USD",
      "location": "example.beancount:70",
      "type": "Assets:Cash",
      "warning": "Unknown directive at line 70, column 29",
    },
  },
  {
    "body": "Expenses:Other   1.00 USD",
    "date": "",
    "kind": "unknown_directive",
    "meta": {
      "body": "Expenses:Other   1.00 USD",
      "location": "example.beancount:71",
      "type": "Expenses:Other",
      "warning": "Unknown directive at line 71, column 28",
    },
  },
  {
    "date": "2015-04-20",
    "flag": "*",
    "kind": "transaction",
    "links": [
      "test",
    ],
    "meta": undefined,
    "narration": "Descrição com "aspas" e 
 nova linha",
    "payee": "Aspas e caracteres",
    "postings": [
      {
        "account": "Assets:Cash",
        "amount": {
          "currency": "USD",
          "value": -5,
        },
        "meta": undefined,
      },
      {
        "account": "Expenses:Other",
        "amount": {
          "currency": "USD",
          "value": 5,
        },
        "meta": undefined,
      },
    ],
    "tags": [
      "escape",
    ],
  },
]
`;
